File: 2017/06_MemoryReallocation/test_memory.lua
-- ======================================================================
-- MemoryReallocation
--   Advent of Code 2017 Day 06 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                    t e s t _ m e m o r y . l u a
-- ======================================================================
-- Test solver for Advent of Code 2017 day 06, MemoryReallocation

-- ----------------------------------------------------------------------
--                                                                require
-- ----------------------------------------------------------------------
local luaunit = require('luaunit')

local memory = require('memory')

-- ----------------------------------------------------------------------
--                                                              from_text
-- ----------------------------------------------------------------------

function from_text(text)
  -- Break the text into trimed, non-comment lines

  -- 1. We start with no lines
  local result = {}

  -- 2. Loop for lines in the text
  for line in text:gmatch('[^\r\n]+') do

    -- 3. But ignore blank and comment lines
    line = line:gsub("%s*$", "")
    if #line > 0 and "!" ~= line:sub(1, 1) then

      -- 4. Add the line
      table.insert(result, line)
    end
  end

  -- 5. Return a table of cleaned lines
  return result
end

-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------
local EXAMPLE_TEXT = [[
0 2 7 0
]]
local PART_ONE_TEXT = EXAMPLE_TEXT
local PART_TWO_TEXT = EXAMPLE_TEXT

local PART_ONE_RESULT = 5
local PART_TWO_RESULT = 4

-- ======================================================================
--                                                             TestMemory
-- ======================================================================

function test_empty_init()
  -- Test the default Memory creation

  -- 1. Create default Memory object
  local myobj = memory:Memory()

  -- 2. Make sure it has the default values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(#myobj.text, 0)
  luaunit.assertEquals(#myobj.banks, 0)
  luaunit.assertEquals(#myobj.history, 0)

end

function test_text_init()
  -- Test the Memory object creation from text

  -- 1. Create Memory object from text
  local myobj = memory:Memory({text=from_text(EXAMPLE_TEXT)})

  -- 2. Make sure it has the expected values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(#myobj.text, 1)
  luaunit.assertEquals(#myobj.banks, 4)
  luaunit.assertEquals(#myobj.history, 1)
  luaunit.assertEquals(myobj.history[1], "0,2,7,0")

-- 3. Check methods
  luaunit.assertEquals(myobj:most_blocks(), 3)
  luaunit.assertEquals(myobj:next_bank(3), 4)
  luaunit.assertEquals(myobj:next_bank(4), 1)
  
  myobj:redistribute(3)
  luaunit.assertEquals(#myobj.history, 2)
  luaunit.assertEquals(myobj.history[2], "2,4,1,2")
  luaunit.assertEquals(myobj:deja_vu(), 2)
  
  luaunit.assertEquals(myobj:most_blocks(), 2)
  myobj:redistribute(2)
  luaunit.assertEquals(#myobj.history, 3)
  luaunit.assertEquals(myobj.history[3], "3,1,2,3")
  luaunit.assertEquals(myobj:deja_vu(), 3)
  
  luaunit.assertEquals(myobj:most_blocks(), 1)
  myobj:redistribute(1)
  luaunit.assertEquals(#myobj.history, 4)
  luaunit.assertEquals(myobj.history[4], "0,2,3,4")
  luaunit.assertEquals(myobj:deja_vu(), 4)
  
  luaunit.assertEquals(myobj:most_blocks(), 4)
  myobj:redistribute(4)
  luaunit.assertEquals(#myobj.history, 5)
  luaunit.assertEquals(myobj.history[5], "1,3,4,1")
  luaunit.assertEquals(myobj:deja_vu(), 5)
  
  luaunit.assertEquals(myobj:most_blocks(), 3)
  myobj:redistribute(3)
  luaunit.assertEquals(#myobj.history, 6)
  luaunit.assertEquals(myobj.history[6], "2,4,1,2")
  luaunit.assertEquals(myobj:deja_vu(), 2)

end

function test_part_one()
  -- Test part one example of Memory object

  -- 1. Create Memory object from text
  local myobj = memory:Memory({text=from_text(PART_ONE_TEXT)})

  -- 2. Check the part one result
  luaunit.assertEquals(myobj:part_one({verbose=false, limit=0}), PART_ONE_RESULT)

end

function test_part_two()
  -- Test part two example of Memory object

  -- 1. Create Memory object from text
  local myobj = memory:Memory({part2=true, text=from_text(PART_TWO_TEXT)})

  -- 2. Check the part two result
  luaunit.assertEquals(myobj:part_two({verbose=false, limit=0}), PART_TWO_RESULT)

end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------
os.exit( luaunit.LuaUnit.run() )

-- ======================================================================
-- end                  t e s t _ m e m o r y . l u a                 end
-- ======================================================================

File: 2024/18_RAMRun/test_memory.py

# ======================================================================
# RAM Run
#   Advent of Code 2024 Day 18 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                    t e s t _ m e m o r y . p y
# ======================================================================
"Test solver for Advent of Code 2024 day 18, RAM Run"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import unittest

import aoc_18
import memory

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
EXAMPLE_TEXT = """
5,4
4,2
4,5
3,0
2,1
6,3
2,4
1,5
0,6
3,3
2,6
5,1
1,2
5,5
2,5
6,5
1,4
0,4
6,4
1,1
6,1
1,0
0,5
1,6
2,0
"""
PART_ONE_TEXT = EXAMPLE_TEXT
PART_TWO_TEXT = EXAMPLE_TEXT

PART_ONE_RESULT = 22
PART_TWO_RESULT = "6,1"

# ======================================================================
#                                                             TestMemory
# ======================================================================


class TestMemory(unittest.TestCase):  # pylint: disable=R0904
    "Test Memory object"

    def test_empty_init(self):
        "Test the default Memory creation"

        # 1. Create default Memory object
        myobj = memory.Memory()

        # 2. Make sure it has the default values
        self.assertEqual(myobj.part2, False)
        self.assertEqual(myobj.text, None)
        self.assertEqual(myobj.size, 70)
        self.assertEqual(myobj.start, (0, 0))
        self.assertEqual(myobj.exit, (70, 70))
        self.assertEqual(myobj.falling, [])
        self.assertEqual(myobj.fall, 1024)
        self.assertEqual(myobj.corrupted, set())

    def test_text_init(self):
        "Test the Memory object creation from text"

        # 1. Create Memory object from text
        myobj = memory.Memory(text=aoc_18.from_text(EXAMPLE_TEXT),
                              size=6, fall=12)

        # 2. Make sure it has the expected values
        self.assertEqual(myobj.part2, False)
        self.assertEqual(len(myobj.text), 25)
        self.assertEqual(myobj.size, 6)
        self.assertEqual(myobj.start, (0, 0))
        self.assertEqual(myobj.exit, (6, 6))
        self.assertEqual(len(myobj.falling), 25)
        self.assertEqual(myobj.fall, 12)
        self.assertEqual(len(myobj.corrupted), 12)

        # 3. Check methods
        self.assertSetEqual(set(myobj.next_step((0, 0))),
                            set([(1, 0), (0, 1)]))
        self.assertSetEqual(set(myobj.next_step((2, 0))),
                            set([(1, 0)]))
        self.assertEqual(myobj.min_steps_to_exit(), 22)
        self.assertEqual(myobj.find_blockage(), (6, 1))

    def test_part_one(self):
        "Test part one example of Memory object"

        # 1. Create Memory object from text
        text = aoc_18.from_text(PART_ONE_TEXT)
        myobj = memory.Memory(text=text, size=6, fall=12)

        # 2. Check the part one result
        self.assertEqual(myobj.part_one(verbose=False), PART_ONE_RESULT)

    def test_part_two(self):
        "Test part two example of Memory object"

        # 1. Create Memory object from text
        text = aoc_18.from_text(PART_TWO_TEXT)
        myobj = memory.Memory(part2=True, text=text, size=6, fall=12)

        # 2. Check the part two result
        self.assertEqual(myobj.part_two(verbose=False), PART_TWO_RESULT)


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------


if __name__ == '__main__':
    pass

# ======================================================================
# end                   t e s t _ m e m o r y . p y                  end
# ======================================================================
